test/e2e/hybrid_test.go

```diff
@@ -163,4 +185,367 @@ var _ = Describe("Hybrid Gateway", func() {
+				},
+				map[string]*v1.Matcher{
+					"sni-star": {
+						SslConfig: &ssl.SslConfig{
+							SniDomains: []string{"*.test.com"},
+						},
+						SourcePrefixRanges: []*v3.CidrRange{
+							{
+								AddressPrefix: "1.2.3.4",
+								PrefixLen: &wrappers.UInt32Value{
+									Value: 32,
+								},
+							},
+						},
+					},
+				}, "sni-star"),
+			Entry("sni match, ip non-match",
+				ClientConnectionProperties{
+					SrcIp: net.ParseIP("1.2.3.4"),
+					SNI:   "foo.test.com",
+				},
+				map[string]*v1.Matcher{
+					"sni-star": {
+						SslConfig: &ssl.SslConfig{
+							SniDomains: []string{"*.test.com"},
+						},
+						SourcePrefixRanges: []*v3.CidrRange{
+							{
+								AddressPrefix: "2.2.3.4",
+								PrefixLen: &wrappers.UInt32Value{
+									Value: 32,
+								},
+							},
+						},
+					},
+				}, NoMatch),
+			Entry("most specific sni matcher",
+				ClientConnectionProperties{
+					SrcIp: net.ParseIP("1.2.3.4"),
+					SNI:   "foo.test.com",
+				},
+				map[string]*v1.Matcher{
+					"less-specific": {
+						SslConfig: &ssl.SslConfig{
+							SniDomains: []string{"*.test.com"},
+						},
+					},
+					"more-specific": {
+						SslConfig: &ssl.SslConfig{
+							SniDomains: []string{"foo.test.com", "*.com"},
+						},
+					},
+				}, "more-specific"),
+			Entry("most specific sni matcher with invalid source ip",
+				// envoy parses sni domain first - it matches 'more-specific',
+				// and then ignores all other matchers at the same level (ie.
+				// the 'less-specific' matcher). as envoy descends through the
+				// 'more-specific' branch, it finds no matching
+				// SourcePrefixRanges values, so it returns no filter chain
+				// found
+				ClientConnectionProperties{
+					SrcIp: net.ParseIP("1.2.3.4"),
+					SNI:   "foo.test.com",
+				},
+				map[string]*v1.Matcher{
+					"less-specific": {
+						SslConfig: &ssl.SslConfig{
+							SniDomains: []string{"*.test.com"},
+						},
+					},
+					"more-specific": {
+						SslConfig: &ssl.SslConfig{
+							SniDomains: []string{"foo.test.com"},
+						},
+						SourcePrefixRanges: []*v3.CidrRange{
+							{
+								AddressPrefix: "2.2.3.4",
+								PrefixLen: &wrappers.UInt32Value{
+									Value: 32,
+								},
+							},
+						},
+					},
+				}, NoMatch),
+			Entry("sni matcher with multiple source ip",
+				ClientConnectionProperties{
+					SrcIp: net.ParseIP("1.2.3.4"),
+					SNI:   "foo.test.com",
+				},
+				map[string]*v1.Matcher{
+					"more-specific": {
+						SslConfig: &ssl.SslConfig{
+							SniDomains: []string{"foo.test.com"},
+						},
+						SourcePrefixRanges: []*v3.CidrRange{
+							{
+								AddressPrefix: "2.2.3.4",
+								PrefixLen: &wrappers.UInt32Value{
+									Value: 32,
+								},
+							},
+							{
+								AddressPrefix: "1.2.3.4",
+								PrefixLen: &wrappers.UInt32Value{
+									Value: 32,
+								},
+							},
+						},
+					},
+				}, "more-specific"),
+			Entry("sni matcher with multiple source ip",
```

Comment:
This test has the same name as the previous

Not sure if this is the best name for it, just a suggestion, feel free to come up with a better one
```suggestion
			Entry("sni matcher with multiple source ip, less precise CIDR range match",
```